envsource.go

```diff
@@ -0,0 +1,465 @@
+package staert
+
+import (
+	"fmt"
+	"os"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/containous/flaeg"
+	"github.com/containous/flaeg/parse"
+	"github.com/fatih/camelcase"
+)
+
+// Loader interface
+// A Loader is an object that can be used to load a configuration from a
+// configuration structure
+type Loader interface {
+	LoadConfig(config interface{}) error
+}
+
+// SourceLoader : can be both a staert.Source and a staert.Loader
+type SourceLoader interface {
+	Loader
+	Source
+}
+
+// envSource implements SourceLoader
+// Enables to populate configuration struct with informations extracted from
+// process's environment variables. Variables names are like %PREFIX%%SEP%%FIELD_NAME%
+// It supports pointer to values and struct, however not slices and arrays..
+type envSource struct {
+	prefix    string
+	separator string
+	parsers   map[reflect.Type]parse.Parser
+}
+
+// NewEnvSource constructs a new instance of envSource
+func NewEnvSource(prefix, separator string, parsers map[reflect.Type]parse.Parser) SourceLoader {
+	return &envSource{prefix, separator, parsers}
+}
+
+// Parse parse and load config structure
+func (e *envSource) Parse(cmd *flaeg.Command) (*flaeg.Command, error) {
+	return cmd, e.LoadConfig(cmd.Config)
+}
+
+func (e *envSource) LoadConfig(config interface{}) error {
+	configVal := reflect.ValueOf(config).Elem()
+
+	values, err := e.analyzeStruct(configVal.Type(), []string{})
+
+	if err != nil {
+		return err
+	}
+
+	return e.assignValues(configVal, values, []string{})
+}
+
+type envValue struct {
+	StrValue string
+	Path     path
+}
+
+type path []string
+
+func (p path) clone() []string {
+	res := make([]string, len(p))
+	copy(res, p)
+	return res
+}
+
+// Recursively scan the given config structure type information
+// and look for defined environment variables.
+func (e *envSource) analyzeStruct(configType reflect.Type, currentPath path) ([]*envValue, error) {
+	res := []*envValue{}
+
+	for i := 0; i < configType.NumField(); i++ {
+		field := configType.Field(i)
+
+		//TODO: Handle this case;
+		//find the underlying struct and process it.
+		if field.Type.Kind() == reflect.Interface {
+			continue //skip fields of kind interface
+		}
+		// If we're facing an embedded struct
+		if field.Anonymous {
+			values, err := e.analyzeValue(field.Type, currentPath)
+
+			if err != nil {
+				return []*envValue{}, err
+			}
+
+			res = append(res, values...)
+			continue
+		}
+
+		// unexported fields must be handled after embdedded structs (field.Anonymous)
+		// because the PkgPath is also null for them.
+		// ref: https://github.com/golang/go/issues/21122
+		if field.PkgPath != "" { //field is unexported
+			continue
+		}
+
+		values, err := e.analyzeValue(field.Type, append(currentPath, field.Name))
+
+		if err != nil {
+			return []*envValue{}, err
+		}
```

Comment:
```go
		values, err := e.analyzeValue(field.Type, append(currentPath, field.Name))
		if err != nil {
			return nil, err
		}
```

Reply:
What is prefered? return nil, or return the an empty array?
I tend to prefer returning an empty array to avoid the caller checking for nil all the time, or forgetting to do so...